File: merger/merger.go
package merger

import (
	"fmt"

	"github.com/xraph/farp"
)

// Merger handles OpenAPI schema composition
type Merger struct {
	config MergerConfig
}

// MergerConfig configures the merger behavior
type MergerConfig struct {
	// Default conflict strategy if not specified in metadata
	DefaultConflictStrategy farp.ConflictStrategy

	// Title for the merged OpenAPI spec
	MergedTitle string

	// Description for the merged OpenAPI spec
	MergedDescription string

	// Version for the merged OpenAPI spec
	MergedVersion string

	// Whether to include service tags in operations
	IncludeServiceTags bool

	// Whether to sort merged content alphabetically
	SortOutput bool

	// Custom server URLs for the merged spec
	Servers []Server
}

// DefaultMergerConfig returns default merger configuration
func DefaultMergerConfig() MergerConfig {
	return MergerConfig{
		DefaultConflictStrategy: farp.ConflictStrategyPrefix,
		MergedTitle:             "Federated API",
		MergedDescription:       "Merged API specification from multiple services",
		MergedVersion:           "1.0.0",
		IncludeServiceTags:      true,
		SortOutput:              true,
		Servers:                 []Server{},
	}
}

// NewMerger creates a new OpenAPI merger
func NewMerger(config MergerConfig) *Merger {
	return &Merger{
		config: config,
	}
}

// MergeResult contains the merged OpenAPI spec and metadata
type MergeResult struct {
	// The merged OpenAPI specification
	Spec *OpenAPISpec

	// Services that were included in the merge
	IncludedServices []string

	// Services that were excluded (not marked for inclusion)
	ExcludedServices []string

	// Conflicts that were encountered during merge
	Conflicts []Conflict

	// Warnings (non-fatal issues)
	Warnings []string
}

// Conflict represents a conflict encountered during merging
type Conflict struct {
	// Type of conflict (path, component, tag, etc.)
	Type ConflictType

	// Path or name that conflicted
	Item string

	// Services involved in the conflict
	Services []string

	// How the conflict was resolved
	Resolution string

	// Conflict strategy that was applied
	Strategy farp.ConflictStrategy
}

// ConflictType represents the type of conflict
type ConflictType string

const (
	// ConflictTypePath indicates path conflict
	ConflictTypePath ConflictType = "path"

	// ConflictTypeComponent indicates component name conflict
	ConflictTypeComponent ConflictType = "component"

	// ConflictTypeTag indicates tag conflict
	ConflictTypeTag ConflictType = "tag"

	// ConflictTypeOperationID indicates operation ID conflict
	ConflictTypeOperationID ConflictType = "operationId"

	// ConflictTypeSecurityScheme indicates security scheme conflict
	ConflictTypeSecurityScheme ConflictType = "securityScheme"
)

// Merge merges multiple OpenAPI schemas from service manifests
func (m *Merger) Merge(schemas []ServiceSchema) (*MergeResult, error) {
	result := &MergeResult{
		Spec: &OpenAPISpec{
			OpenAPI: "3.1.0", // Use latest version
			Info: Info{
				Title:       m.config.MergedTitle,
				Description: m.config.MergedDescription,
				Version:     m.config.MergedVersion,
			},
			Servers: m.config.Servers,
			Paths:   make(map[string]PathItem),
			Components: &Components{
				Schemas:         make(map[string]map[string]interface{}),
				Responses:       make(map[string]Response),
				Parameters:      make(map[string]Parameter),
				RequestBodies:   make(map[string]RequestBody),
				SecuritySchemes: make(map[string]SecurityScheme),
			},
			Tags:       []Tag{},
			Extensions: make(map[string]interface{}),
		},
		IncludedServices: []string{},
		ExcludedServices: []string{},
		Conflicts:        []Conflict{},
		Warnings:         []string{},
	}

	// Track what we've seen for conflict detection
	seenPaths := make(map[string]string)           // path -> service
	seenComponents := make(map[string]string)      // component -> service
	seenOperationIDs := make(map[string]string)    // operationID -> service
	seenTags := make(map[string]Tag)               // tag name -> tag
	seenSecuritySchemes := make(map[string]string) // security scheme -> service

	// Process each schema
	for _, schema := range schemas {
		serviceName := schema.Manifest.ServiceName

		// Check if this schema should be included
		if !shouldIncludeInMerge(schema) {
			result.ExcludedServices = append(result.ExcludedServices, serviceName)
			continue
		}

		result.IncludedServices = append(result.IncludedServices, serviceName)

		// Parse the schema if not already parsed
		if schema.Parsed == nil {
			parsed, err := ParseOpenAPISchema(schema.Schema)
			if err != nil {
				result.Warnings = append(result.Warnings,
					fmt.Sprintf("Failed to parse schema for %s: %v", serviceName, err))
				continue
			}
			schema.Parsed = parsed
		}

		// Get composition config
		compConfig := getCompositionConfig(schema.Manifest)
		strategy := m.getConflictStrategy(compConfig)

		// Determine prefixes
		componentPrefix := getComponentPrefix(schema.Manifest, compConfig)
		tagPrefix := getTagPrefix(schema.Manifest, compConfig)
		operationIDPrefix := getOperationIDPrefix(schema.Manifest, compConfig)

		// Merge paths
		paths := ApplyRouting(schema.Parsed.Paths, schema.Manifest)
		for path, pathItem := range paths {
			// Check for path conflicts
			if existingService, exists := seenPaths[path]; exists {
				conflict := Conflict{
					Type:     ConflictTypePath,
					Item:     path,
					Services: []string{existingService, serviceName},
					Strategy: strategy,
				}

				switch strategy {
				case farp.ConflictStrategyError:
					return nil, fmt.Errorf("path conflict: %s exists in both %s and %s",
						path, existingService, serviceName)

				case farp.ConflictStrategySkip:
					conflict.Resolution = fmt.Sprintf("Skipped path from %s", serviceName)
					result.Conflicts = append(result.Conflicts, conflict)
					continue

				case farp.ConflictStrategyOverwrite:
					conflict.Resolution = fmt.Sprintf("Overwritten with %s version", serviceName)
					result.Conflicts = append(result.Conflicts, conflict)
					// Continue to overwrite

				case farp.ConflictStrategyPrefix:
					// Add service prefix to path
					newPath := fmt.Sprintf("/%s%s", serviceName, path)
					conflict.Resolution = fmt.Sprintf("Prefixed to %s", newPath)
					result.Conflicts = append(result.Conflicts, conflict)
					path = newPath

				case farp.ConflictStrategyMerge:
					// Attempt to merge operations
					pathItem = mergePathItems(result.Spec.Paths[path], pathItem)
					conflict.Resolution = "Merged operations"
					result.Conflicts = append(result.Conflicts, conflict)
				}
			}

			// Apply prefixes to operation IDs and tags
			pathItem = applyOperationPrefixes(pathItem, operationIDPrefix, tagPrefix, serviceName,
				seenOperationIDs, result)

			result.Spec.Paths[path] = pathItem
			seenPaths[path] = serviceName
		}

		// Merge components
		if schema.Parsed.Components != nil {
			prefixedComponents := PrefixComponentNames(schema.Parsed.Components, componentPrefix)

			for name, schemaObj := range prefixedComponents.Schemas {
				if existingService, exists := seenComponents[name]; exists {
					conflict := Conflict{
						Type:     ConflictTypeComponent,
						Item:     name,
						Services: []string{existingService, serviceName},
						Strategy: strategy,
					}

					if strategy == farp.ConflictStrategySkip {
						conflict.Resolution = fmt.Sprintf("Skipped component from %s", serviceName)
						result.Conflicts = append(result.Conflicts, conflict)
						continue
					}

					conflict.Resolution = fmt.Sprintf("Overwritten with %s version", serviceName)
					result.Conflicts = append(result.Conflicts, conflict)
				}

				result.Spec.Components.Schemas[name] = schemaObj
				seenComponents[name] = serviceName
			}

			// Merge other component types
			for name, response := range prefixedComponents.Responses {
				result.Spec.Components.Responses[name] = response
			}
			for name, param := range prefixedComponents.Parameters {
				result.Spec.Components.Parameters[name] = param
			}
			for name, body := range prefixedComponents.RequestBodies {
				result.Spec.Components.RequestBodies[name] = body
			}

			// Merge security schemes (with conflict detection)
			for name, scheme := range prefixedComponents.SecuritySchemes {
				if existingService, exists := seenSecuritySchemes[name]; exists {
					conflict := Conflict{
						Type:     ConflictTypeSecurityScheme,
						Item:     name,
						Services: []string{existingService, serviceName},
						Strategy: strategy,
					}

					switch strategy {
					case farp.ConflictStrategyError:
						return nil, fmt.Errorf("security scheme conflict: %s exists in both %s and %s",
							name, existingService, serviceName)

					case farp.ConflictStrategySkip:
						conflict.Resolution = fmt.Sprintf("Skipped security scheme from %s", serviceName)
						result.Conflicts = append(result.Conflicts, conflict)
						continue

					case farp.ConflictStrategyOverwrite:
						conflict.Resolution = fmt.Sprintf("Overwritten with %s version", serviceName)
						result.Conflicts = append(result.Conflicts, conflict)

					case farp.ConflictStrategyPrefix:
						// Prefix the security scheme name
						prefixedName := componentPrefix + "_" + name
						conflict.Resolution = fmt.Sprintf("Prefixed to %s", prefixedName)
						result.Conflicts = append(result.Conflicts, conflict)
						result.Spec.Components.SecuritySchemes[prefixedName] = scheme
						seenSecuritySchemes[prefixedName] = serviceName
						continue

					case farp.ConflictStrategyMerge:
						// For security schemes, merge is same as overwrite with warning
						conflict.Resolution = fmt.Sprintf("Merged (overwritten) with %s version", serviceName)
						result.Conflicts = append(result.Conflicts, conflict)
					}
				}

				result.Spec.Components.SecuritySchemes[name] = scheme
				seenSecuritySchemes[name] = serviceName
			}
		}

		// Merge tags
		for _, tag := range schema.Parsed.Tags {
			if tagPrefix != "" && m.config.IncludeServiceTags {
				tag.Name = tagPrefix + "_" + tag.Name
			}

			if existing, exists := seenTags[tag.Name]; exists {
				// Merge descriptions
				if tag.Description != "" && existing.Description == "" {
					existing.Description = tag.Description
					seenTags[tag.Name] = existing
				}
			} else {
				seenTags[tag.Name] = tag
				result.Spec.Tags = append(result.Spec.Tags, tag)
			}
		}
	}

	// Sort output if requested
	if m.config.SortOutput {
		result.Spec.Tags = SortTags(result.Spec.Tags)
	}

	return result, nil
}

// Helper functions

func shouldIncludeInMerge(schema ServiceSchema) bool {
	// Check OpenAPI metadata for composition config
	for _, schemaDesc := range schema.Manifest.Schemas {
		if schemaDesc.Type == farp.SchemaTypeOpenAPI &&
			schemaDesc.Metadata != nil &&
			schemaDesc.Metadata.OpenAPI != nil &&
			schemaDesc.Metadata.OpenAPI.Composition != nil {
			return schemaDesc.Metadata.OpenAPI.Composition.IncludeInMerged
		}
	}

	// Default: include if OpenAPI schema is present
	for _, schemaDesc := range schema.Manifest.Schemas {
		if schemaDesc.Type == farp.SchemaTypeOpenAPI {
			return true
		}
	}

	return false
}

func getCompositionConfig(manifest *farp.SchemaManifest) *farp.CompositionConfig {
	for _, schemaDesc := range manifest.Schemas {
		if schemaDesc.Type == farp.SchemaTypeOpenAPI &&
			schemaDesc.Metadata != nil &&
			schemaDesc.Metadata.OpenAPI != nil {
			return schemaDesc.Metadata.OpenAPI.Composition
		}
	}
	return nil
}

func (m *Merger) getConflictStrategy(config *farp.CompositionConfig) farp.ConflictStrategy {
	if config != nil && config.ConflictStrategy != "" {
		return config.ConflictStrategy
	}
	return m.config.DefaultConflictStrategy
}

func getComponentPrefix(manifest *farp.SchemaManifest, config *farp.CompositionConfig) string {
	if config != nil && config.ComponentPrefix != "" {
		return config.ComponentPrefix
	}
	return manifest.ServiceName
}

func getTagPrefix(manifest *farp.SchemaManifest, config *farp.CompositionConfig) string {
	if config != nil && config.TagPrefix != "" {
		return config.TagPrefix
	}
	return manifest.ServiceName
}

func getOperationIDPrefix(manifest *farp.SchemaManifest, config *farp.CompositionConfig) string {
	if config != nil && config.OperationIDPrefix != "" {
		return config.OperationIDPrefix
	}
	return manifest.ServiceName
}

func applyOperationPrefixes(item PathItem, opIDPrefix, tagPrefix, serviceName string,
	seenOperationIDs map[string]string, result *MergeResult,
) PathItem {
	applyToOp := func(op *Operation) {
		if op == nil {
			return
		}

		// Prefix operation ID
		if op.OperationID != "" {
			originalID := op.OperationID
			if opIDPrefix != "" {
				op.OperationID = opIDPrefix + "_" + op.OperationID
			}

			// Check for conflicts
			if existingService, exists := seenOperationIDs[op.OperationID]; exists {
				result.Conflicts = append(result.Conflicts, Conflict{
					Type:       ConflictTypeOperationID,
					Item:       originalID,
					Services:   []string{existingService, serviceName},
					Resolution: fmt.Sprintf("Prefixed to %s", op.OperationID),
				})
			}
			seenOperationIDs[op.OperationID] = serviceName
		}

		// Prefix tags
		if tagPrefix != "" {
			op.Tags = PrefixTags(op.Tags, tagPrefix)
		}
	}

	applyToOp(item.Get)
	applyToOp(item.Post)
	applyToOp(item.Put)
	applyToOp(item.Delete)
	applyToOp(item.Patch)
	applyToOp(item.Options)
	applyToOp(item.Head)
	applyToOp(item.Trace)

	return item
}

func mergePathItems(existing, new PathItem) PathItem {
	// Merge operations - prefer non-nil operations
	if new.Get != nil {
		existing.Get = new.Get
	}
	if new.Post != nil {
		existing.Post = new.Post
	}
	if new.Put != nil {
		existing.Put = new.Put
	}
	if new.Delete != nil {
		existing.Delete = new.Delete
	}
	if new.Patch != nil {
		existing.Patch = new.Patch
	}
	if new.Options != nil {
		existing.Options = new.Options
	}
	if new.Head != nil {
		existing.Head = new.Head
	}
	if new.Trace != nil {
		existing.Trace = new.Trace
	}

	return existing
}

File: farp-ts/src/merger/merger.ts
/**
 * OpenAPI schema merger for FARP TypeScript client.
 *
 * Ported from Go: merger/merger.go
 * Handles merging multiple OpenAPI schemas from service manifests into a
 * single unified specification with conflict detection and resolution.
 */

import type {
  SchemaManifest,
  CompositionConfig,
  ConflictStrategy,
} from '../types';
import type {
  OpenAPISpec,
  PathItem,
  Components,
  Tag,
  Conflict,
  MergeResult,
  ServiceSchema,
  Server,
} from './openapi-types';
import {
  prefixTags,
  sortTags,
  applyRouting,
  prefixComponentNames,
  parseOpenAPISchema,
  applyOperationPrefixes,
} from './utils';

// =============================================================================
// MergerConfig
// =============================================================================

/** Configuration for the Merger. */
export interface MergerConfig {
  /** Default conflict strategy if not specified in metadata */
  defaultConflictStrategy?: ConflictStrategy;

  /** Title for the merged OpenAPI spec */
  mergedTitle?: string;

  /** Description for the merged OpenAPI spec */
  mergedDescription?: string;

  /** Version for the merged OpenAPI spec */
  mergedVersion?: string;

  /** Whether to include service tags in operations */
  includeServiceTags?: boolean;

  /**
   * When true, all operations from a service are grouped under a single
   * tag matching the service name, instead of prefixing individual tags.
   * Takes precedence over includeServiceTags when both are true.
   */
  collapseServiceTags?: boolean;

  /** Whether to sort merged content alphabetically */
  sortOutput?: boolean;

  /** Custom server URLs for the merged spec */
  servers?: Server[];
}

/** Returns default merger configuration. */
export function defaultMergerConfig(): MergerConfig {
  return {
    defaultConflictStrategy: 'prefix',
    mergedTitle: 'Federated API',
    mergedDescription: 'Merged API specification from multiple services',
    mergedVersion: '1.0.0',
    includeServiceTags: true,
    collapseServiceTags: false,
    sortOutput: true,
    servers: [],
  };
}

// =============================================================================
// Merger
// =============================================================================

/** Handles OpenAPI schema composition / merging. */
export class Merger {
  private readonly config: MergerConfig;

  constructor(config: MergerConfig) {
    this.config = config;
  }

  /**
   * Merges multiple OpenAPI schemas from service manifests.
   * Returns the merged spec, included/excluded services, conflicts, and warnings.
   */
  merge(schemas: ServiceSchema[]): MergeResult {
    const result: MergeResult = {
      spec: {
        openapi: '3.1.0',
        info: {
          title: this.config.mergedTitle ?? 'Federated API',
          description: this.config.mergedDescription ?? 'Merged API specification from multiple services',
          version: this.config.mergedVersion ?? '1.0.0',
        },
        servers: this.config.servers ?? [],
        paths: {},
        components: {
          schemas: {},
          responses: {},
          parameters: {},
          requestBodies: {},
          securitySchemes: {},
        },
        tags: [],
        extensions: {},
      },
      includedServices: [],
      excludedServices: [],
      conflicts: [],
      warnings: [],
    };

    // Track what we've seen for conflict detection
    const seenPaths = new Map<string, string>(); // path -> service
    const seenComponents = new Map<string, string>(); // component -> service
    const seenOperationIds = new Map<string, string>(); // operationID -> service
    const seenTags = new Map<string, Tag>(); // tag name -> tag
    const seenSecuritySchemes = new Map<string, string>(); // security scheme -> service

    // Process each schema
    for (let schema of schemas) {
      const serviceName = schema.manifest.service_name;

      // Check if this schema should be included
      if (!shouldIncludeInMerge(schema)) {
        result.excludedServices.push(serviceName);
        continue;
      }

      result.includedServices.push(serviceName);

      // Parse the schema if not already parsed
      if (!schema.parsed) {
        const parsed = parseOpenAPISchema(schema.schema);
        if (!parsed) {
          result.warnings.push(
            `Failed to parse schema for ${serviceName}`,
          );
          continue;
        }
        schema = { ...schema, parsed };
      }

      // Get composition config
      const compConfig = getCompositionConfig(schema.manifest);
      const strategy = this.getConflictStrategy(compConfig);

      // Determine prefixes
      const componentPrefix = getComponentPrefix(schema.manifest, compConfig);
      const tagPrefix = getTagPrefix(schema.manifest, compConfig);
      const operationIdPrefix = getOperationIdPrefix(schema.manifest, compConfig);

      // Merge paths
      const paths = applyRouting(schema.parsed!.paths, schema.manifest);

      for (let [path, pathItem] of Object.entries(paths)) {
        // Check for path conflicts
        const existingService = seenPaths.get(path);
        if (existingService) {
          const conflict: Conflict = {
            type: 'path',
            item: path,
            services: [existingService, serviceName],
            resolution: '',
            strategy,
          };

          switch (strategy) {
            case 'error':
              throw new Error(
                `path conflict: ${path} exists in both ${existingService} and ${serviceName}`,
              );

            case 'skip':
              conflict.resolution = 'Skipped path from ' + serviceName;
              result.conflicts.push(conflict);
              continue;

            case 'overwrite':
              conflict.resolution = `Overwritten with ${serviceName} version`;
              result.conflicts.push(conflict);
              // Continue to overwrite
              break;

            case 'prefix': {
              // Add service prefix to path
              const newPath = `/${serviceName}${path}`;
              conflict.resolution = 'Prefixed to ' + newPath;
              result.conflicts.push(conflict);
              path = newPath;
              break;
            }

            case 'merge':
              // Attempt to merge operations
              pathItem = mergePathItems(result.spec.paths[path], pathItem);
              conflict.resolution = 'Merged operations';
              result.conflicts.push(conflict);
              break;
          }
        }

        // Apply prefixes to operation IDs and tags
        pathItem = applyOperationPrefixes(
          pathItem,
          operationIdPrefix,
          tagPrefix,
          serviceName,
          this.config.collapseServiceTags ?? false,
          seenOperationIds,
          result,
        );

        result.spec.paths[path] = pathItem;
        seenPaths.set(path, serviceName);
      }

      // Merge components
      if (schema.parsed!.components) {
        const prefixedComponents = prefixComponentNames(
          schema.parsed!.components,
          componentPrefix,
        );

        // Merge schemas
        if (prefixedComponents.schemas) {
          for (const [name, schemaObj] of Object.entries(prefixedComponents.schemas)) {
            const existingCompService = seenComponents.get(name);
            if (existingCompService) {
              const conflict: Conflict = {
                type: 'component',
                item: name,
                services: [existingCompService, serviceName],
                resolution: '',
                strategy,
              };

              if (strategy === 'skip') {
                conflict.resolution = 'Skipped component from ' + serviceName;
                result.conflicts.push(conflict);
                continue;
              }

              conflict.resolution = `Overwritten with ${serviceName} version`;
              result.conflicts.push(conflict);
            }

            result.spec.components!.schemas![name] = schemaObj;
            seenComponents.set(name, serviceName);
          }
        }

        // Merge other component types
        if (prefixedComponents.responses) {
          Object.assign(result.spec.components!.responses!, prefixedComponents.responses);
        }
        if (prefixedComponents.parameters) {
          Object.assign(result.spec.components!.parameters!, prefixedComponents.parameters);
        }
        if (prefixedComponents.requestBodies) {
          Object.assign(result.spec.components!.requestBodies!, prefixedComponents.requestBodies);
        }

        // Merge security schemes (with conflict detection)
        if (prefixedComponents.securitySchemes) {
          for (const [name, scheme] of Object.entries(prefixedComponents.securitySchemes)) {
            const existingSecService = seenSecuritySchemes.get(name);
            if (existingSecService) {
              const conflict: Conflict = {
                type: 'securityScheme',
                item: name,
                services: [existingSecService, serviceName],
                resolution: '',
                strategy,
              };

              switch (strategy) {
                case 'error':
                  throw new Error(
                    `security scheme conflict: ${name} exists in both ${existingSecService} and ${serviceName}`,
                  );

                case 'skip':
                  conflict.resolution = 'Skipped security scheme from ' + serviceName;
                  result.conflicts.push(conflict);
                  continue;

                case 'overwrite':
                  conflict.resolution = `Overwritten with ${serviceName} version`;
                  result.conflicts.push(conflict);
                  break;

                case 'prefix': {
                  const prefixedName = `${componentPrefix}_${name}`;
                  conflict.resolution = 'Prefixed to ' + prefixedName;
                  result.conflicts.push(conflict);
                  result.spec.components!.securitySchemes![prefixedName] = scheme;
                  seenSecuritySchemes.set(prefixedName, serviceName);
                  continue;
                }

                case 'merge':
                  conflict.resolution = `Merged (overwritten) with ${serviceName} version`;
                  result.conflicts.push(conflict);
                  break;
              }
            }

            result.spec.components!.securitySchemes![name] = scheme;
            seenSecuritySchemes.set(name, serviceName);
          }
        }
      }

      // Merge tags
      if (this.config.collapseServiceTags) {
        // Collapse all tags into a single service-level tag
        const serviceTag: Tag = {
          name: serviceName,
          description: 'Routes from ' + serviceName,
        };
        if (!seenTags.has(serviceTag.name)) {
          seenTags.set(serviceTag.name, serviceTag);
          result.spec.tags!.push(serviceTag);
        }
      } else {
        const parsedTags = schema.parsed!.tags ?? [];
        for (const tag of parsedTags) {
          let tagName = tag.name;
          if (tagPrefix && this.config.includeServiceTags) {
            tagName = `${tagPrefix}_${tag.name}`;
          }

          const existing = seenTags.get(tagName);
          if (existing) {
            // Merge descriptions
            if (tag.description && !existing.description) {
              existing.description = tag.description;
              seenTags.set(tagName, existing);
            }
          } else {
            const newTag: Tag = { name: tagName, description: tag.description };
            seenTags.set(tagName, newTag);
            result.spec.tags!.push(newTag);
          }
        }
      }
    }

    // Sort output if requested
    if (this.config.sortOutput) {
      result.spec.tags = sortTags(result.spec.tags ?? []);
    }

    return result;
  }

  /** Resolves the conflict strategy from composition config or default. */
  private getConflictStrategy(
    config: CompositionConfig | undefined,
  ): ConflictStrategy {
    if (config && config.conflict_strategy) {
      return config.conflict_strategy;
    }
    return this.config.defaultConflictStrategy ?? 'prefix';
  }
}

// =============================================================================
// Helper functions
// =============================================================================

/**
 * Determines whether a service schema should be included in the merged output.
 */
function shouldIncludeInMerge(schema: ServiceSchema): boolean {
  // Check OpenAPI metadata for composition config
  for (const schemaDesc of schema.manifest.schemas) {
    if (
      schemaDesc.type === 'openapi' &&
      schemaDesc.metadata?.openapi?.composition
    ) {
      return schemaDesc.metadata.openapi.composition.include_in_merged;
    }
  }

  // Default: include if OpenAPI schema is present
  for (const schemaDesc of schema.manifest.schemas) {
    if (schemaDesc.type === 'openapi') {
      return true;
    }
  }

  return false;
}

/**
 * Extracts the composition config from a manifest's OpenAPI schema metadata.
 */
function getCompositionConfig(
  manifest: SchemaManifest,
): CompositionConfig | undefined {
  for (const schemaDesc of manifest.schemas) {
    if (
      schemaDesc.type === 'openapi' &&
      schemaDesc.metadata?.openapi
    ) {
      return schemaDesc.metadata.openapi.composition;
    }
  }
  return undefined;
}

/**
 * Gets the component prefix for a service (from composition config or service name).
 */
function getComponentPrefix(
  manifest: SchemaManifest,
  config: CompositionConfig | undefined,
): string {
  if (config && config.component_prefix) {
    return config.component_prefix;
  }
  return manifest.service_name;
}

/**
 * Gets the tag prefix for a service (from composition config or service name).
 */
function getTagPrefix(
  manifest: SchemaManifest,
  config: CompositionConfig | undefined,
): string {
  if (config && config.tag_prefix) {
    return config.tag_prefix;
  }
  return manifest.service_name;
}

/**
 * Gets the operation ID prefix for a service (from composition config or service name).
 */
function getOperationIdPrefix(
  manifest: SchemaManifest,
  config: CompositionConfig | undefined,
): string {
  if (config && config.operation_id_prefix) {
    return config.operation_id_prefix;
  }
  return manifest.service_name;
}

/**
 * Merges two path items by preferring non-null operations from the new item.
 */
function mergePathItems(existing: PathItem, newPath: PathItem): PathItem {
  const merged = { ...existing };

  if (newPath.get) merged.get = newPath.get;
  if (newPath.post) merged.post = newPath.post;
  if (newPath.put) merged.put = newPath.put;
  if (newPath.delete) merged.delete = newPath.delete;
  if (newPath.patch) merged.patch = newPath.patch;
  if (newPath.options) merged.options = newPath.options;
  if (newPath.head) merged.head = newPath.head;
  if (newPath.trace) merged.trace = newPath.trace;

  return merged;
}
